code/modules/mob/living/carbon/xenomorph/abilities.dm

```diff
@@ -1371,76 +1362,154 @@ GLOBAL_LIST_INIT(pattern_images_list, list(
+		else
+			move_hologram(target_turf, holograms[index])
+			check_turf_validity(target_turf, holograms[index])
+		index++
+	start_cleanup_timer()
+
+/datum/action/ability/activable/xeno/place_pattern/proc/check_turf_validity(turf/target_turf, obj/effect/hologram)
+	var/datum/action/ability/activable/xeno/secrete_resin/secrete_resin = locate() in xeno_owner.actions
+	if(!secrete_resin)
+		return
+	hologram.remove_filter("invalid_turf_filter")
+	if(!secrete_resin.can_build_here(target_turf, TRUE))
+		hologram.add_filter("invalid_turf_filter", 1, color_matrix_filter(rgb(233, 23, 23)))
+
+/datum/action/ability/activable/xeno/place_pattern/proc/should_replace_holograms(list/target_turfs)
+	if(length(holograms) != length(target_turfs))
+		return TRUE
+	var/index = 1
+	for(var/obj/hologram as anything in holograms)
+		if(get_dist(hologram, target_turfs[index]) > move_range)
+			return TRUE
+	index++
+	return FALSE
+
+/datum/action/ability/activable/xeno/place_pattern/proc/create_hologram(turf/target_turf)
+	var/atom/selected = xeno_owner.selected_resin
+	var/obj/effect/build_hologram/hologram = new(target_turf, selected)
+	hologram.alpha = 0
+	hologram.layer = selected.layer + 1
+	hologram.step_size = move_range * ICON_SIZE_ALL
+	animate(hologram, 1 SECONDS, alpha = initial(hologram.alpha))
+	holograms += hologram
+
+/datum/action/ability/activable/xeno/place_pattern/proc/move_hologram(turf/target_turf, obj/effect/hologram)
+	hologram.abstract_move(target_turf)
+
+/datum/action/ability/activable/xeno/place_pattern/proc/start_cleanup_timer()
+	delete_timer()
+	cleanup_timer = addtimer(CALLBACK(src, PROC_REF(cleanup_holograms)), cleanup_time, TIMER_STOPPABLE)
+
+/datum/action/ability/activable/xeno/place_pattern/proc/delete_timer()
+	deltimer(cleanup_timer)
+	cleanup_timer = null
+
+/datum/action/ability/activable/xeno/place_pattern/proc/cleanup_holograms()
+	delete_timer()
+	QDEL_LIST(holograms)
 
 ///Selects the pattern from a radial menu
-/datum/action/ability/activable/xeno/place_pattern/proc/selectpattern()
+/datum/action/ability/activable/xeno/place_pattern/proc/select_pattern()
 	var/pattern_choice = show_radial_menu(owner, owner, GLOB.pattern_images_list, radius = 48)
 	if(!pattern_choice)
 		return
 	selected_pattern = new pattern_choice
 	var/datum/buildingpattern/pattern = pattern_choice
+	var/image/pattern_icon = GLOB.pattern_images_list[pattern_choice]
+	action_icon_state = pattern_icon.icon_state
+	update_button_icon()
 	xeno_owner.balloon_alert(xeno_owner, pattern.overheadmsg)
+	cleanup_holograms()
 
 /datum/action/ability/activable/xeno/place_pattern/use_ability(atom/A)
+	var/datum/action/ability/activable/xeno/secrete_resin/secrete_resin = locate() in xeno_owner.actions
+	if(!istype(secrete_resin))
+		to_chat(xeno_owner, span_warning("We need to be able to secrete resin to use [src]!"))
+		return FALSE
+	var/list/target_turfs = get_target_turfs(A)
+	if(!length(target_turfs))
+		to_chat(xeno_owner, span_warning("We can't build here!"))
```

Comment:
could be a balloon alert

Reply:
this is a very edge case anyway, but yea fine